File: macos/Gutenkern/Sources/GutenkernCore/MarkHistory.swift
import Foundation

public final class MarkHistory {
    private var undoStack: [KerningMarkState] = []
    private var redoStack: [KerningMarkState] = []
    private var coalescing = false
    private var coalesced = false

    public init() {}

    public var canUndo: Bool { !undoStack.isEmpty }
    public var canRedo: Bool { !redoStack.isEmpty }

    public func record(from previous: KerningMarkState, to next: KerningMarkState) {
        guard previous != next else {
            return
        }
        if coalescing {
            if !coalesced {
                undoStack.append(previous)
                redoStack.removeAll()
                coalesced = true
            }
            return
        }
        undoStack.append(previous)
        redoStack.removeAll()
    }

    public func beginCoalescing() {
        coalescing = true
        coalesced = false
    }

    public func endCoalescing() {
        coalescing = false
        coalesced = false
    }

    public func undo(current: KerningMarkState) -> KerningMarkState? {
        guard let previous = undoStack.popLast() else {
            return nil
        }
        redoStack.append(current)
        return previous
    }

    public func redo(current: KerningMarkState) -> KerningMarkState? {
        guard let next = redoStack.popLast() else {
            return nil
        }
        undoStack.append(current)
        return next
    }
}

File: windows/Gutenkern.Core/MarkHistory.cs
namespace Gutenkern;

public sealed class MarkHistory
{
    private readonly List<KerningMarkState> _undo = [];
    private readonly List<KerningMarkState> _redo = [];
    private bool _coalescing;
    private bool _coalesced;

    public bool CanUndo => _undo.Count > 0;
    public bool CanRedo => _redo.Count > 0;

    public void Record(KerningMarkState previous, KerningMarkState next)
    {
        if (Same(previous, next))
        {
            return;
        }

        if (_coalescing)
        {
            if (!_coalesced)
            {
                _undo.Add(Copy(previous));
                _redo.Clear();
                _coalesced = true;
            }

            return;
        }

        _undo.Add(Copy(previous));
        _redo.Clear();
    }

    public void BeginCoalescing()
    {
        _coalescing = true;
        _coalesced = false;
    }

    public void EndCoalescing()
    {
        _coalescing = false;
        _coalesced = false;
    }

    public KerningMarkState? Undo(KerningMarkState current)
    {
        if (_undo.Count == 0)
        {
            return null;
        }

        var previous = _undo[^1];
        _undo.RemoveAt(_undo.Count - 1);
        _redo.Add(Copy(current));
        return previous;
    }

    public KerningMarkState? Redo(KerningMarkState current)
    {
        if (_redo.Count == 0)
        {
            return null;
        }

        var next = _redo[^1];
        _redo.RemoveAt(_redo.Count - 1);
        _undo.Add(Copy(current));
        return next;
    }

    private static bool Same(KerningMarkState left, KerningMarkState right) =>
        left.Done.SetEquals(right.Done);

    private static KerningMarkState Copy(KerningMarkState state) =>
        new(state.Done);
}
